### ISSUE TITLE
Unittest compilation of std.string crash (Windows): too deep recursion

### ISSUE THREAD

**Issue by JohanEngelen** [ldc-developers/ldc#1297]
On Windows, this stops compilation (`-w`!) without any diagnostic:
`C:\projects\ninja-ldc\bin\ldc2.exe --output-o -c -IC:/projects/ldc/runtime/druntime/src -IC:/projects/ldc/runtime/druntime/src/gc C:/projects/ldc/runtime/phobos/std/string.d -ofC:/projects/ninja-ldc/runtime/std/string-unittest-debug.obj -w -d -link-debuglib -unittest -IC:/projects/ldc/runtime/phobos`

**Comment by JohanEngelen** [comment]
(Perhaps the bug is not directly related to `-w`)

**Comment by kinke** [comment]
On merge-2.069 only, right? I'll investigate once I'm home (in 6+ hours).

**Comment by JohanEngelen** [comment]
yeah.
I thought I had isolated the issue, reduced it to a test case, and solved it by removing `-w` for now, but... somewhere I went wrong and the issue is still there.
Disabling the unittests up to line 1200 or so will fix compilation.

**Comment by JohanEngelen** [comment]
Perhaps it is an out-of-memory issue? I am able to turn the crash on/off by duplicating/removing the asserts around line 1220.

**Comment by JohanEngelen** [comment]
~~~Perhaps it is an out-of-memory issue?~~~
Nope, with a loop, I've been able to get LDC to allocate more than 10GB of memory, no problem.

**Comment by JohanEngelen** [comment]
Looks like an infinite recursion of `TemplateInstance.needsCodeGen()`

**Comment by JohanEngelen** [comment]
I think I have found the problem:  `TemplateInstance.needsCodeGen()` is a recursive function. As it turns out, for some of the unittests of std.string (e.g. line 1222) the recursion depth is very very deep. The call from declarations.cpp, `void visit(TemplateInstance *decl)`, reaches a recursion depth that is too much on Windows and the program simply exits.
I don't know how to solve this, other than rewriting needsCodeGen to something non-recursive; I think it can be turned into a loop.
What do you think?

**Comment by JohanEngelen** [comment]
(in one case, the depth is >10000)

**Comment by JohanEngelen** [comment]
@9rnsr Do you think it is possible to rewrite `TemplateInstance.needsCodeGen()` into something non-recursive? Or at least a refactoring that reduces the depth of recursion?

**Comment by kinke** [comment]
Oh wow. That most likely also expains why we're hitting a rare LLVM bug (https://github.com/ldc-developers/ldc/issues/930) when compiling the `std.string` unittests with debug infos on Win64.

**Comment by yebblies** [comment]
We do explicitly set the stack size for DMD on windows, since the default is too small.  What is it set to in LDC?

**Comment by JohanEngelen** [comment]
Thanks, found it: https://github.com/D-Programming-Language/dmd/blob/master/src/win32.mak#L285
I'll try that.
Is there a maximum nested template instantiation depth? (LDC has an experimental feature for setting that)

**Comment by kinke** [comment]
Apparently the default stack size used by MSVC is a single MB. Linux uses 8, and that's what DMD chose on Windows as well. Let's go with the same size.

**Comment by yebblies** [comment]
Yes, it has nested template instantiation limits but they are set up for the 8MB stack.  We use 8MB for dmd because I looked up the linux default and went with that.

**Comment by JohanEngelen** [comment]
Mac also uses 8MB as default.

**Comment by JohanEngelen** [comment]
Fixed in merge-2.069 branch
